cpp/src/join/key_remapping.cu

```diff
@@ -0,0 +1,651 @@
+  {
+    return cudf::bit_is_set(_row_bitmask, i);
+  }
+
+ private:
+  cudf::bitmask_type const* _row_bitmask;
+};
+
+/**
+ * @brief Device functor to create a pair of {hash_value, row_index}.
+ */
+template <typename T, typename Hasher>
+class make_key_pair {
+ public:
+  CUDF_HOST_DEVICE constexpr make_key_pair(Hasher const& hash) : _hash{hash} {}
+
+  __device__ __forceinline__ auto operator()(cudf::size_type i) const noexcept
+  {
+    return cuco::pair{_hash(i), T{i}};
+  }
+
+ private:
+  Hasher _hash;
+};
+
+/**
+ * @brief Device output transform functor to extract row index from hash table result.
+ */
+struct extract_index {
+  __device__ constexpr cudf::size_type operator()(
+    cuco::pair<hash_value_type, rhs_index_type> const& x) const
+  {
+    return static_cast<cudf::size_type>(x.second);
+  }
+};
+
+/**
+ * @brief Comparator adapter for probe-time comparison (two-table).
+ */
+template <typename Equal>
+struct probe_comparator {
+  probe_comparator(Equal const& d_equal) : _d_equal{d_equal} {}
+
+  __device__ constexpr auto operator()(
+    cuco::pair<hash_value_type, lhs_index_type> const& lhs,
+    cuco::pair<hash_value_type, rhs_index_type> const& rhs) const noexcept
+  {
+    if (lhs.first != rhs.first) { return false; }
+    return _d_equal(lhs.second, rhs.second);
+  }
+
+ private:
+  Equal _d_equal;
+};
+
+/**
+ * @brief Comparator adapter for primitive row operators (probe-time).
+ */
+template <typename Equal>
+struct primitive_probe_comparator {
+  primitive_probe_comparator(Equal const& d_equal) : _d_equal{d_equal} {}
+
+  __device__ constexpr auto operator()(
+    cuco::pair<hash_value_type, lhs_index_type> const& lhs,
+    cuco::pair<hash_value_type, rhs_index_type> const& rhs) const noexcept
+  {
+    if (lhs.first != rhs.first) { return false; }
+    return _d_equal(static_cast<cudf::size_type>(lhs.second),
+                    static_cast<cudf::size_type>(rhs.second));
+  }
+
+ private:
+  Equal _d_equal;
+};
+
+/**
+ * @brief Comparator adapter for build-time self-comparison (deduplication).
+ */
+template <typename RowEqual>
+struct build_comparator {
+  build_comparator(RowEqual const& d_equal) : _d_equal{d_equal} {}
+
+  __device__ constexpr auto operator()(
+    cuco::pair<hash_value_type, rhs_index_type> const& lhs,
+    cuco::pair<hash_value_type, rhs_index_type> const& rhs) const noexcept
+  {
+    if (lhs.first != rhs.first) { return false; }
+    return _d_equal(static_cast<cudf::size_type>(lhs.second),
+                    static_cast<cudf::size_type>(rhs.second));
+  }
+
+ private:
+  RowEqual _d_equal;
+};
+
+/**
+ * @brief Functor to check if a count is greater than zero.
+ */
+struct is_nonzero_count {
+  __device__ bool operator()(cudf::size_type c) const { return c > 0; }
+};
+
+/**
+ * @brief Functor for inserting keys with counting (with null filtering).
+ */
+template <typename SetRef, typename KeyIter>
+struct insert_and_count_with_nulls_fn {
```

Comment:
Indeed the functor `insert_and_count_fn` below is unnecessary. We can just use this same implementation for both cases (with nulls and non-nulls). For the non-nulls cases, just construct the functor with a `nullptr` value for the bitmask. Then, modify the statement:
```
if (!cudf::bit_is_set(bitmask_ptr, idx)) { return; }
```
into this:
```
if (bitmask_ptr && !cudf::bit_is_set(bitmask_ptr, idx)) { return; }
```